### DISCUSSION TITLE
Convergence when using ScalarLagrangeMultiplier

### DISCUSSION THREAD

**Discussion by charlotte-b8086** [idaholab/moose#27336] [Q&A General]
I am working to code up a Poisson-Nernst-Planck model by [Bazant et al.](https://web.mit.edu/bazant/www/papers/pdf/Bazant_2005_SIAM_J_Appl_Math_MTF2.pdf):

```tex
\begin{aligned}
\nabla c_{+} + c_{+} \nabla \phi &= 4j \\
\nabla c_{-} - c_{-} \nabla \phi &= 0 \\
-\epsilon^2 \Delta \phi &= \frac{1}{2}\left(c_{+} -c_{-}\right) \notag
\end{aligned}
subject to the boundary conditions
\begin{aligned*}
\phi(0) &= 0 \\
\phi(1) &= 1 \\
c_{+}(0) &= 1.09 \\
c_{+}(1) &= 0.91
\end{aligned*}
and the constraint 
$$
\int_{\Omega} c_{-} dx = 1.
$$
```
 In the paper, the authors integrated two second order differential equations to obtain the first order differential equations (1) and (2) using a boundary condition for $c_{-}$. I have substituted some values directly into the model that appears in the paper to simplify the presentation. 

I have successfully been able to run a MOOSE input file which solves for the system of PDEs with the boundary conditions but without the constraint.

I am now working to implement the constraint using ScalarLagrangeMultiplier. It seems like a straightforward application of the available kernel. 

However, as soon as I introduce this constraint into the input file, the solver struggles to converge. 
I started with just coding $c_{-}$ with the constraint (letting all the other variables be 0 essentially) and that was able to converge. I was able to add in $\phi$ with just its diffusion term (so working with the simplified PDE $-\epsilon^2 \Delta \phi = 0$), but as soon as I add in the CoupledForce term for phi  (so working with the simplified PDE $-\epsilon^2 \Delta \phi = -0.5 c_{-}$), the solver is unable to converge. 

Looking through documentation, it was mentioned somewhere that adding in constraints does make solving more difficult.

Do you have any recommendations for what type of Executioner and/or Preconditioning would help convergence? 

This is what I currently have, which works for just the diffusion term but not the CoupledForce term in the residual equation for $\phi$.

```c++
[Preconditioning]
  # Preconditioning is required for Newton's method.
  [./coupled]
    type = SMP
    full = true
    solve_type = 'NEWTON'
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'LINEAR'
  l_max_its = 30
  l_tol = 1e-6
  nl_max_its = 300
  nl_abs_tol = 1e-4
  #line_search = none
  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type
                         -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'lu      31                  preonly
                         lu          1'
[]
``` 
I was working with the LU in the petsc_options_value because when I first introduced the diffusion term (so did not just set $\phi$ to be 0), I was having a zero factor pivot error, and that solution was proposed in another discussion post. 

I believe it was before I added in $\phi$ I was having a line_search error issue and saw in another discussion post that sometimes the default does not work for convergence, so I had input the line_search = none to rectify that, but then I believe when I added in $\phi$, while I had to add in the LU, I was able to get rid of that command without convergence failing.

I'm trying to recreate the results of the paper so I won't really know if the solution I am getting is correct until everything is incorporated. Even though I have convergence for this simple case, I don't know that it is to the correct solution, but I want to have convergence in the more complicated case so that I can verify that it is correct.

I appreciate any suggestions you may offer.

**Comment by GiudGiud** [comment]
Hello

Have you tried to add a nonzero factor shift? Often constraints DoFs introduce a 0 on the diagonal 

What kind of convergence error are you getting? The linear or nonlinear solve?

Guillaume

**Reply by charlotte-b8086** [reply]
This helped a lot, thank you! My code is now running (although not yet producing there correct results). It was linear solve convergence.
